### PR TITLE
Envvar assignment cleanup

### PR DESCRIPTION
**Pull request by keszybz**
I noticed a minor memleak while working on a patch for the manager. This resulted in a small cleanup series.

### CODE REVIEW

File: src/basic/env-util.c [hunk 1]
```diff
@@ -395,6 +394,24 @@ int strv_env_replace(char ***l, char *p) {
...
+        char *p;
+        int r;
+
+        /* Like strv_env_replace(), but copies the argument. */
+
+        p = strdup(assignment);
+        if (!p)
+                return -ENOMEM;
+
+        r = strv_env_replace(l, p);
```
Comment (yuwata) [inline src/basic/env-util.c]: It may retrun `-EINVAL` when the input string does not contain `=`, and in that case, `p` is not consumed...

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/basic/env-util.c (hunk 1 above)
Comment (keszybz) [reply src/basic/env-util.c]: Duh, I even put a TAKE_PTR there, but forgot about `_cleanup_`. Fixed now.
(And also the unused var flagged by clang.)

File: src/basic/env-util.c [hunk 2]
```diff
@@ -370,9 +370,8 @@ int strv_env_replace(char ***l, char *p) {
 
         assert(p);
 
-        /* Replace first occurrence of the env var or add a new one in the string list. Drop other occurrences. Edits
-         * in-place. Does not copy p.  p must be a valid key=value assignment.
-         */
+        /* Replace first occurrence of the env var or add a new one in the string list. Drop other
+         * occurrences. Edits in-place. Does not copy p. p must be a valid key=value assignment. */
```
Comment (poettering) [inline src/basic/env-util.c]: the fact that `p` is donated is weird... maybe we should rename the function strv_env_replace_donate() or so, and then make strv_env_replace() the function that strdup()s.... i.e it's the exception that we don't copy the string, we should make that clearer I think

or, at the least, insert a comment right into the argument list `/* donated! */` or so

File: src/locale/keymap-util.c [hunk 3]
```diff
@@ -66,9 +66,7 @@ static void context_free_vconsole(Context *c) {
 }
 
 static void context_free_locale(Context *c) {
-        int p;
-
-        for (p = 0; p < _VARIABLE_LC_MAX; p++)
+        for (int p = 0; p < _VARIABLE_LC_MAX; p++)
```
Comment (poettering) [inline src/locale/keymap-util.c]: int → LocaleVariable?

File: src/locale/keymap-util.c [hunk 4]
```diff
@@ -85,9 +83,7 @@ void context_clear(Context *c) {
 };
 
 void locale_simplify(char *locale[_VARIABLE_LC_MAX]) {
-        int p;
-
-        for (p = VARIABLE_LANG+1; p < _VARIABLE_LC_MAX; p++)
+        for (int p = VARIABLE_LANG+1; p < _VARIABLE_LC_MAX; p++)
```
Comment (poettering) [inline src/locale/keymap-util.c]: ditto

File: src/locale/keymap-util.c [hunk 5]
```diff
@@ -138,13 +134,11 @@ int locale_read_data(Context *c, sd_bus_message *m) {
...
                         return r;
         } else {
-                int p;
-
                 c->locale_mtime = USEC_INFINITY;
                 context_free_locale(c);
 
                 /* Fill in what we got passed from systemd. */
-                for (p = 0; p < _VARIABLE_LC_MAX; p++) {
+                for (int p = 0; p < _VARIABLE_LC_MAX; p++) {
```
Comment (poettering) [inline src/locale/keymap-util.c]: ditto

File: src/locale/keymap-util.c [hunk 6]
```diff
@@ -294,11 +288,11 @@ int x11_read_data(Context *c, sd_bus_message *m) {
 int locale_write_data(Context *c, char ***settings) {
         _cleanup_strv_free_ char **l = NULL;
         struct stat st;
-        int r, p;
+        int r;
 
         /* Set values will be returned as strv in *settings on success. */
 
-        for (p = 0; p < _VARIABLE_LC_MAX; p++) {
+        for (int p = 0; p < _VARIABLE_LC_MAX; p++) {
```
Comment (poettering) [inline src/locale/keymap-util.c]: ditto

File: src/basic/env-util.c [hunk 7]
```diff
@@ -406,7 +406,30 @@ int strv_env_replace_strdup(char ***l, const char *assignment) {
         r = strv_env_replace(l, p);
         if (r < 0)
                 return r;
+        TAKE_PTR(p);
+        return r;
+}
+
+int strv_env_assign(char ***l, const char *key, const char *value) {
```
Comment (poettering) [inline src/basic/env-util.c:411]: the commit msg is botched

File: src/basic/env-util.c (hunk 2 above)
Comment (keszybz) [reply src/basic/env-util.c]: Right, that's a good point. I added a commit to rework strv_env_replace() into strv_env_replace_consume(). It makes almost all callers quite a bit simpler.

**Comment by yuwata** [comment]
LGTM.

**Comment by poettering** [comment]
lgtm, too

**Pull request merged**